Add listener/pool cipher allow list

BenjaminLudwigSAP opened the pull request:
- Add unit tests for listener/pool ciphers allow list
- Add `tls_cipher_allow_list` config option
- If the allow list is not set (`None`), only the prohibit list is enforced
- If there is a cipher conflict, tell the API user the cipher allow list, if it is set
- Fix `check_cipher_prohibit_list`: If prohibit list is empty, return immediately. Else there will be a false positive cipher conflict if `cipherstring.split(':')` contains `''`.
- Rename `check_default_ciphers_prohibit_list_conflict` to `check_default_ciphers_conflict`
- Extend `check_default_ciphers_conflict` to check for ciphers not in allow list.

BenjaminLudwigSAP committed:
Add failing unit tests for listener/pool ciphers allow list

BenjaminLudwigSAP committed:
Enforce listener/pool cipher allow list

- Add tls_cipher_allow_list config option

- If the allow list is not set (None), only the prohibit list is enforced

- If there is a cipher conflict, tell the API user the cipher allow
  list, if it is set

- Fix check_cipher_prohibit_list: If cipherstring is empty, return
  immediately. Else ''.split(':') will be [''], with '' matching the
  prohibit list, if it is empty, creating a false positive conflict.

- Rename check_default_ciphers_prohibit_list_conflict to
  check_default_ciphers_conflict

- Extend check_default_ciphers_conflict to check for ciphers not in
  allow list.

notandy commented on octavia/common/validate.py:
```diff
@@ -440,21 +442,49 @@ def check_cipher_prohibit_list(cipherstring):
 
-def check_default_ciphers_prohibit_list_conflict():
+def check_cipher_allow_list(cipherstring):
+    if cipherstring == '' or CONF.api_settings.tls_cipher_allow_list is None:
```
cipherstring == '' check seems to be unintended copy, according to the option, only unset (==none) will disable the allow check. Even if unlikely, allow_list='' should disallow all ciphers.

BenjaminLudwigSAP replied:
The `cipherstring == ''` is the aforementioned fix ported to `check_cipher_allow_list`, but you're right. It isn't needed if I simply check with `not CONF.api_settings.tls_cipher_allow_list`, which I was about to do anyway :)

BenjaminLudwigSAP replied:
The reason behind having the `cipherstring == ''` check was the aforementioned fix:
> Fix check_cipher_prohibit_list: If cipherstring is empty, return immediately. Else ''.split(':') will be [''], with '' matching the prohibit list, if it is empty, creating a false positive conflict.

1) This made `test_check_default_ciphers_allow_list_conflict` run green before functionality implementation, even though it should run red at that point.
2) If the prohibit list is empty (`''`), Octavia users can provoke a conflict by setting `tls_ciphers` to `:` or something that includes `::`, so that the list resulting from the split includes `''`, which matches the empty prohibit list. This doesn't make sense semantically. However, the error message will tell the user `The following ciphers have been prohibited by an administrator: , ` or something similar.

However, I implemented the check wrong. We only need to return if the prohibit/allow list is empty.
I'll append a fix commit.

notandy reviewed (changes requested):
Only one minor issue, but rest looks good!

notandy commented on octavia/common/validate.py:
```diff
@@ -431,6 +431,8 @@ def ip_not_reserved(ip_address):
 
 
 def check_cipher_prohibit_list(cipherstring):
+    if cipherstring == '':
```
Seems unnecessary

BenjaminLudwigSAP replied:
See comment in other thread.

BenjaminLudwigSAP committed:
Fix: Prevent false positive cipher conflicts

False positive cipher conflicts in check_cipher_prohibit_list can
occur if one of the following cases is true:
- cipherstring == ''
- cipherstring.split(':').contains('')
- CONF.api_settings.tls_cipher_prohibit_list is None
- CONF.api_settings.tls_cipher_prohibit_list == ''
So far I only checked the first case. However, the last two cases are
sufficient to return early, since the cipherstring doesn't matter
then.
So check only them.

Similarly in check_cipher_allow_list.
